receiver/kafkareceiver/consumer_franz.go

```diff
@@ -0,0 +1,475 @@
+func (c *franzConsumer) consume(ctx context.Context, size int) bool {
+	fetch := c.client.PollRecords(ctx, size)
+	defer c.client.AllowRebalance()
+
+	if err := fetch.Err0(); fetch.IsClientClosed() {
+		c.settings.Logger.Info("consumer stopped", zap.Error(err))
+		return false // Shut down the consumer loop.
+	}
+	// There's a variety of errors that are returned by fetch.Errors(). We
+	// handle the errors that require a client restart above. The rest can
+	// simply be logged and keep fetching.
+	var hasError bool
+	fetch.EachError(func(topic string, partition int32, err error) {
+		c.settings.Logger.Error("consumer fetch error", zap.Error(err),
+			zap.String(attrTopic, topic),
+			zap.String(attrPartition, strconv.Itoa(int(partition))),
+		)
+		if !hasError {
+			hasError = true
+		}
+	})
+	if hasError || fetch.Empty() {
+		return true // Return right away after errors or empty fetch.
+	}
+
+	select { // Check if Shutdown's been called.
+	case <-ctx.Done():
+		return false
+	case <-c.closing:
+		// The client is now closed (Shutdown has been called), so we need to
+		// return and stop consuming. At this point it's safe to return, since
+		// no offsets have been marked for commits (autocommit) or committed.
+		return false
+	default:
+	}
+
+	// Acquire the read lock on each consume to ensure the client is not closed
+	// and the assignments map is not modified while consuming. Copy the map
+	// to avoid locking for the duration of the consume loop.
+	c.mu.RLock()
+	assignments := make(map[topicPartition]*pc, len(c.assignments))
+	maps.Copy(assignments, c.assignments)
+	c.mu.RUnlock()
+
+	var wg sync.WaitGroup
+	// Process messages on a per partition basis, wait for them to finish and
+	// commit the processed records (if autocommit is disabled).
+	fetch.EachPartition(func(p kgo.FetchTopicPartition) {
+		count := len(p.Records)
+		if count == 0 {
+			return // Skip partitions without any records.
+		}
+		tp := topicPartition{topic: p.Topic, partition: p.Partition}
+		assign, ok := assignments[tp]
+		// NOTE(marclop): This could happen if the partition is lost between
+		// the time the assignments map is copied and the partition is accessed.
+		if !ok {
+			c.settings.Logger.Warn(
+				"attempted to process records for a partition not assigned to this consumer",
+				zap.String(attrTopic, tp.topic),
+				zap.String(attrPartition, strconv.Itoa(int(tp.partition))),
+			)
+			return
+		}
+		wg.Add(1)
+		assign.logger.Debug("processing fetched records",
+			zap.Int("count", count),
+			zap.Int64("start_offset", p.Records[0].Offset),
+			zap.Int64("end_offset", p.Records[count-1].Offset),
+		)
+		go func(pc *pc, msgs []*kgo.Record) {
+			defer wg.Done()
+			fatalOffset := int64(-1)
+			var lastProcessed *kgo.Record
+			for _, msg := range msgs {
+				if !c.config.MessageMarking.After {
+					c.client.MarkCommitRecords(msg)
+				}
+				m := wrapFranzMsg(msg)
+				if err := c.handleMessage(pc.ctx, m, pc); err != nil {
```

Comment:
Does pc.ctx get cancelled on shutdown? Would be good to have a test that ensures Shutdown interrupts a blocking operation in the next consumer, called by handleMessage. (If we don't already)

Reply:
Fixed and added a test.